### PR TITLE
[comms,commsdsl,comms-ublox] update ports

### PR DESCRIPTION
**Pull request by mathisloge**
**Describe the pull request**

- #### What does your PR fix?  

* updates comms to 3.4
* adds feature tools to comms. 
* update commsdsl to 3.6.4
* update comms-ublox to 0.20.2

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
yes

### CODE REVIEW

**Comment by mathisloge** [comment]
\ping https://github.com/commschamp/comms_champion/issues/13

File: ports/comms/portfile.cmake [hunk 1]
```diff
@@ -1,26 +1,60 @@
...
 
+vcpkg_check_features(
+    OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+        tools   COMMS_BUILD_TOOLS
+)
+
+if(COMMS_BUILD_TOOLS)
+    set(VCPKG_POLICY_DLLS_WITHOUT_LIBS enabled)
+endif()
```
Comment (JackBoosY) [inline ports/comms/portfile.cmake:17]: Why do that?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/comms/portfile.cmake (hunk 1 above)
Comment (mathisloge) [reply ports/comms/portfile.cmake]: good catch. Forgot to remove that.

File: ports/comms/portfile.cmake [hunk 2]
```diff
@@ -1,26 +1,56 @@
...
+    file(COPY "${CURRENT_PACKAGES_DIR}/lib/CommsChampion/plugin" DESTINATION "${CURRENT_PACKAGES_DIR}/tools/${PORT}/lib/CommsChampion/")
+endif()
+
+vcpkg_fixup_cmake_targets(CONFIG_PATH "lib/LibComms/cmake" TARGET_PATH "share/LibComms")
+
+if(COMMS_BUILD_TOOLS)
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib/CommsChampion/plugin")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib/LibComms")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/lib/CommsChampion/plugin")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/lib/LibComms")
```
Comment (JackBoosY) [inline ports/comms/portfile.cmake:46]: Can you confirm which file was installed in those directories?

File: ports/comms/portfile.cmake [hunk 3]
```diff
@@ -1,26 +1,56 @@
...
-file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib")
-file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
+
+if(COMMS_BUILD_TOOLS)
+    vcpkg_copy_tools(
+        TOOL_NAMES cc_dump cc_view
+        SEARCH_DIR "${CURRENT_PACKAGES_DIR}/bin"
+        AUTO_CLEAN
+    )
+    file(COPY "${CURRENT_PACKAGES_DIR}/lib/CommsChampion/plugin" DESTINATION "${CURRENT_PACKAGES_DIR}/tools/${PORT}/lib/CommsChampion/")
```
Comment (JackBoosY) [inline ports/comms/portfile.cmake:37]: Maybe `RENAME` would be better.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/comms/portfile.cmake (hunk 2 above)
Comment (mathisloge) [reply ports/comms/portfile.cmake]: with the fix now, all dirs are empty (except plugins which are installed now)

File: ports/comms/portfile.cmake (hunk 3 above)
Comment (mathisloge) [reply ports/comms/portfile.cmake]: Rename gives some strange errors that the target dir does not exists

**Comment by JackBoosY** [comment]
Depends on https://github.com/microsoft/vcpkg/pull/19629.

File: ports/comms/portfile.cmake [hunk 4]
```diff
@@ -1,26 +1,57 @@
...
+
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib/LibComms")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib/CommsChampion")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/lib/LibComms")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/lib/CommsChampion")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+else()
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
```
Comment (JackBoosY) [inline ports/comms/portfile.cmake:52]: Why did you remove that? Is this port a header only library?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/comms/portfile.cmake (hunk 4 above)
Comment (mathisloge) [reply ports/comms/portfile.cmake:52]: Yes

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
When building `comms[core,tools]:x64-windows-static`:
```
-- Installing: G:/19581/vcpkg/packages/comms_x64-windows-static/share/comms/copyright
-- Performing post-build validation
DLLs should not be present in a static build, but the following DLLs were found:

    G:/19581/vcpkg/packages/comms_x64-windows-static/bin/comms_champion.dll
    G:/19581/vcpkg/packages/comms_x64-windows-static/debug/bin/comms_champion.dll

There should be no bin\ directory in a static build, but G:\19581\vcpkg\packages\comms_x64-windows-static\bin is present.
There should be no debug\bin\ directory in a static build, but G:\19581\vcpkg\packages\comms_x64-windows-static\debug/bin is present.
If the creation of bin\ and/or debug\bin\ cannot be disabled, use this in the portfile to remove them

    if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
        file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
    endif()

Found 2 error(s). Please correct the portfile:
    G:\19581\vcpkg\ports\comms\portfile.cmake
-- Performing post-build validation done
Error: Building package comms:x64-windows-static failed with: POST_BUILD_CHECKS_FAILED
```

**Comment by mathisloge** [comment]
@JackBoosY Is there a way to add `supports` fields to a feature property? 
The tool itself builds a shared library (hardcoded) which is needed for the build plugins. 
So the feature tools can only be build if the triplet is dynamic, but the port without the feature tools supports all triplets, since it is header only

File: ports/comms/portfile.cmake [hunk 5]
```diff
@@ -1,26 +1,62 @@
...
+# check before configure
+if(COMMS_BUILD_TOOLS)
+    vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY) 
+endif()
+
 vcpkg_configure_cmake(
-    SOURCE_PATH ${SOURCE_PATH}
+    SOURCE_PATH "${SOURCE_PATH}"
     PREFER_NINJA
     OPTIONS
```
Comment (JackBoosY) [inline ports/comms/portfile.cmake:23]:
```suggestion
    OPTIONS ${FEATURE_OPTIONS}
```

File: ports/comms/portfile.cmake [hunk 6]
```diff
@@ -1,26 +1,62 @@
...
+vcpkg_check_features(
+    OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+        tools   COMMS_BUILD_TOOLS
+)
+
+# check before configure
+if(COMMS_BUILD_TOOLS)
+    vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY) 
+endif()
```
Comment (JackBoosY) [inline ports/comms/portfile.cmake:18]: I don't think we can deprive users of the power to build static, even with feature tools.
Please replace `vcpkg_check_linkage` with `message(FATAL_ERROR)` here.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by dg0yt** [comment]
Is the DLL only used by the tools (= a private implementation detail)?
If yes, can't the DLL and its import library be removed from everything but `tools/${PORT}`?

**Comment by mathisloge** [comment]
> Is the DLL only used by the tools (= a private implementation detail)?
> If yes, can't the DLL and its import library be removed from everything but `tools/${PORT}`?

It's not a private implementation detail. It can be used to develop own plugins for the tool application.

File: ports/comms/portfile.cmake (hunk 5 above)
Comment (mathisloge) [reply ports/comms/portfile.cmake:23]: not applicable in this case. There is not a `COMMS_BUILD_TOOLS` in the upstream cmake. 
The tools feature depends on multiple variables. 
```
-DCC_BUILD_TOOLS_LIBRARY=${COMMS_BUILD_TOOLS}
-DCC_INSTALL_TOOLS_LIBRARY=${COMMS_BUILD_TOOLS}
-DCC_BUILD_TOOLS=${COMMS_BUILD_TOOLS}
-DCC_INSTALL_TOOLS=${COMMS_BUILD_TOOLS}
-DCC_BUILD_DEMO_PROTOCOL=OFF
```

Comment (dg0yt) [reply ports/comms/portfile.cmake:23]: But you can set multiple variables from the same feature, even inverted ones.

Comment (mathisloge) [reply ports/comms/portfile.cmake:25]: Ah thanks for the hint. wasn't aware of it

**Comment by mathisloge** [comment]
@JackBoosY please review again. I've also replaced the deprecated cmake functions with the new one

File: ports/comms/portfile.cmake (hunk 5 above)
Comment (mathisloge) [reply ports/comms/portfile.cmake:25]: Ah thanks! wasn't aware of it

File: ports/comms/portfile.cmake [hunk 7]
```diff
@@ -1,26 +1,61 @@
...
+    OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+        tools   CC_BUILD_TOOLS_LIBRARY
+        tools   CC_INSTALL_TOOLS_LIBRARY
+        tools   CC_BUILD_TOOLS
+        tools   CC_INSTALL_TOOLS
+)
+
+# check before configure
+if(CC_BUILD_TOOLS)
```
Comment (JackBoosY) [inline ports/comms/portfile.cmake:19]:
```suggestion
if("tools" IN_LIST FEATURES)
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Review by NancyLi1013 — APPROVED** [review]

**Comment by ras0219-msft** [comment]
This all looks clean and reasonable. Thanks for the update!

**Pull request merged**